Remove usage of Kernel::processor in program.cpp

ruizhangTT opened the pull request:
### Ticket
#26820
#26826

### What's changed
- Check processor clash by HAL descriptor rather than RISCV value.
- Export HAL descriptor in kernel_bins_transfer_info rather than RISCV values.
- Also changed data_collection.cpp to accept HAL descriptor as parameter.

Changes to Kernel API:
- `expected_num_binaries` is moved from `KernelImpl` to `kernel`, because
  1. `get_binary_packed_size` is already exposed, but currently there is no way to know the valid range of its `index` parameter without arch-specific knowledge or `KernelImpl`.
  2. Same reason for the new `get_kernel_processor_type` API.
- Added `get_kernel_processor_class` and `get_kernel_processor_type`. These together with `get_kernel_programmable_core_type` are the processor identifier as defined by HAL, and function as a replacement of `processor`.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16997422399) CI running
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16997424786) CI passed

ruizhangTT committed:
Remove usage of Kernel::processor in program.cpp

- Check processor clash by HAL descriptor rather than RISCV value.
- Export HAL descriptor in kernel_bins_transfer_info rather than RISCV values.
- Also changed data_collection.cpp to accept HAL descriptor as parameter.

Changes to Kernel API:
- `expected_num_binaries` is moved from KernelImpl to kernel, because
  1. `get_binary_packed_size` is already exposed, but currently there is no way
     to know the valid range of its `index` parameter without arch-specific
     knowledge or KernelImpl.
  2. Same reason for the new `get_kernel_processor_type` API.
- Added `get_kernel_processor_class` and `get_kernel_processor_type`.
  These together with `get_kernel_programmable_core_type` are the processor
  identifier as defined by HAL, and function as a replacement of `processor`.

ruizhangTT committed:
Small changes

ruizhangTT committed:
Changes to data_collection

ruizhangTT committed:
Add comment

ruizhangTT committed:
Merge branch 'main' into ruizhang/kernel_riscv

ruizhangTT commented on tt_metal/impl/kernels/kernel.cpp:
```diff
@@ -245,6 +247,21 @@ const std::vector<const ll_api::memory*>& KernelImpl::binaries(uint32_t build_ke
+
+uint32_t ComputeKernel::get_kernel_processor_type(int index) const {
+    TT_ASSERT(0 <= index && index < expected_num_binaries(), "index out of bounds");
+    return index;
```
In HAL, processor type is an int, there is no `enum ProcessorType...`

ruizhangTT commented on tt_metal/impl/dispatch/data_collection.cpp:
```diff
@@ -88,19 +97,20 @@ class DispatchData {
             if (type == DISPATCH_DATA_BINARY) {
-                outfile << "\t  " << riscv_and_data.first << " binary data:\n";
-                riscv_stats.Dump(outfile, riscv_and_data.second);
+                TT_ASSERT(processor != std::nullopt);
```
This is indeed not ideal, but it reflects current usage of data_collection - DISPATCH_DATA_BINARY is only used with riscv != MAX (same as processor != nullopt), and others are only used with riscv == MAX.

omilyutin-tt commented on tt_metal/impl/dispatch/data_collection.cpp:
```diff
@@ -49,7 +57,7 @@ class DispatchStats {
     }
 
-    void Dump(std::ofstream& outfile, std::map<uint32_t, uint32_t>& raw_data) {
+    void Dump(std::ofstream& outfile, const std::map<uint32_t, uint32_t>& raw_data) const {
```
👍

omilyutin-tt commented on tt_metal/impl/dispatch/data_collection.cpp:
```diff
@@ -66,19 +74,20 @@ class DispatchStats {
-        for (auto& riscv_and_data : other.data) {
-            for (auto& size_and_count : riscv_and_data.second) {
-                this->data[riscv_and_data.first][size_and_count.first] += size_and_count.second;
+        for (auto& [processor, processor_data] : other.data) {
+            for (auto& [size, count] : processor_data) {
```
So much better

omilyutin-tt commented on tt_metal/llrt/hal.hpp:
```diff
@@ -32,6 +33,11 @@ enum class ARCH;
+// Tuple of core type, processor class, and processor type to unique identify any processor.
+using HalProcessorIdentifier = std::tuple<HalProgrammableCoreType, HalProcessorClassType, int>;
+// Compile-time maximum for processor types count for any arch.  Useful for creating bitsets.
+static constexpr int MAX_PROCESSOR_TYPES_COUNT = 3;
```
is it for `HalProcessorClassType`? possible to use the enum library to get the same counter?

ruizhangTT replied:
This is for the third component, which is >= 0 and < `get_processor_types_count` from HAL.

omilyutin-tt commented on tt_metal/impl/dispatch/data_collection.cpp:
```diff
@@ -112,7 +122,8 @@ class DispatchData {
 
 private:
-    std::map<RISCV, std::map<uint32_t, uint32_t>> data;  // RISCV -> transaction size -> count
+    // processor -> transaction size -> count
+    std::map<std::optional<HalProcessorIdentifier>, std::map<uint32_t, uint32_t>> data;
```
Do we need a map here? Are keys for `HalProcessorIdentifier` known upfront?

Same for `transaction size -> count`. Strange that unordered map doesn't suffice here

ruizhangTT replied:
Oh I think I know why the old code used std::map in several places - its output depends on the iteration order, and it wanted to output things in order.  Perhaps it doesn't have to use std::map everywhere.

For this particular line, it wants to output in RISCV order (now HalProcessorIdentifier order), and for each entry, it wants to output in key order (transaction size order).

So I'll keep it as-is.

ruizhangTT commented on tt_metal/impl/dispatch/data_collection.cpp:
```diff
@@ -27,6 +30,11 @@ using namespace tt::tt_metal;
 
+std::ostream& operator<<(std::ostream& os, const HalProcessorIdentifier& id) {
+    using enchantum::iostream_operators::operator<<;
+    return os << std::get<0>(id) << "_" << std::get<1>(id) << "_" << std::get<2>(id);
```
I thought about this, but I'm not sure if this formatting is to be accepted everywhere.   If yes, I think this definition can be moved to `hal.cpp`.

omilyutin-tt commented on tt_metal/llrt/hal.hpp:
```diff
@@ -32,6 +33,11 @@ enum class ARCH;
 
 namespace tt_metal {
 
+// Tuple of core type, processor class, and processor type to unique identify any processor.
+using HalProcessorIdentifier = std::tuple<HalProgrammableCoreType, HalProcessorClassType, int>;
```
IIUC not all C++ libs implement hashing for `std::tuple`, does this pass CI? Either way, I think named structs is worth it

ruizhangTT replied:
Changed to struct with named fields.  Implemented `<`, `==` and `std::hash` for it.

ruizhangTT committed:
Address code review comments

- Named fields instead of tuple
- ostream operator<< moved to hal.cpp
- fixed spelling

ruizhangTT commented on tt_metal/llrt/hal.hpp:
```diff
@@ -32,6 +33,11 @@ enum class ARCH;
 
 namespace tt_metal {
 
+// Tuple of core type, processor class, and processor type to unique identify any processor.
```
Fixed, thanks!

omilyutin-tt commented on tt_metal/llrt/hal.hpp:
```diff
@@ -32,6 +34,20 @@ enum class ARCH;
 
+// Struct of core type, processor class, and processor type to uniquely identify any processor.
+struct HalProcessorIdentifier {
+    HalProgrammableCoreType core_type;
+    HalProcessorClassType processor_class;
+    int processor_type;
```
Default initialize the primitive types to avoid UB

ruizhangTT replied:
Done, thanks!

ruizhangTT committed:
Default initialize the primitive types to avoid UB
